src/dmd/dtemplate.d

```diff
@@ -6264,73 +6268,33 @@ extern (C++) class TemplateInstance : ScopeDsymbol
 
         if (global.params.allInst)
         {
-            return true;
```

Comment:
Why was this changed - this isn't related to the unittest hack? IIUC the changes correctly `-allinst` will no longer be a valid workaround for missing template instances caused by instance expected from other (read as non-root) modules.

Reply:
> IIUC the changes correctly -allinst will no longer be a valid workaround

Atila asked me to look into the templates emission stuff because `-allinst` made things worse for one of his earlier experiments - more missing symbols, not less. So the assumption is that it now works as people would expect it to. Before rebasing, I've let the CI testsuite run with an enforced `-allinst`, and it was all green.

Reply:
Elaborating a bit: after removing the `-unittest` special cases for template emission, all hell broke loose, lots of undefined symbols. It took a while to tweak the default (non-`allinst`) template emission algorithm to get rid of most of them; ignoring non-root unittests fixed the last CI failures. My results for druntime/Phobos (+0.5% code size without `-unittest`) show that I had to cull a bit less to get there.

So I hope the situation wrt. spurious missing undefined template symbols (especially for separate per-module compilations) improves with these changes. Coverage is definitely better, because all `-unittest` compiles during CI now test the default template emission algorithm.

---

`-allinst` was supposed to work around these issues, but might make things worse - not just for initial experiments here, but also for https://issues.dlang.org/show_bug.cgi?id=19123, https://issues.dlang.org/show_bug.cgi?id=19406, …

The algorithm I now use for `-allinst` is what I expected it to be - emitting *all* templates instantiated in some root module. So when e.g. compiling 10 modules (to 1 or 10 object files), the set of object files should be self-sufficient wrt. instantiated symbols - not referencing any external templated symbol at all.

Reply:
Thanks for the detailed writeup.

Initially thought about the problem of linker errors introduced by `-checkaction=context`, i.e. templates expected from druntime / phobos which don't include them due to `-release` or `-checkaction=D` (e.g. for LDC with `--link-defaultlib-debug`). `-allinst` used to be a viable workaround but now has no effect - the templates are missing  (even when combined with  `-i` / explicitly passing the module) for `master` *and*  also this patch.

Seems like this patch doesn't make it worse for `-checkaction=context` and improves for everything else.

Reply:
> the templates are missing (even when combined with -i / explicitly passing the module) for master and also this patch

Can you post reproduction steps or file an issue? Thanks!

Reply:
Use an `assert` with the same types as done in any module imported from Druntime / Phobos., e.g. `assert(someInt)`

See [19937](https://issues.dlang.org/show_bug.cgi?id=19937) and [22374](https://issues.dlang.org/show_bug.cgi?id=22374).

Reply:
Thx, can reproduce with issue 22374 and `-allinst` - with this patch, 2 undefined template symbols instantiated inside `_d_assert_fail` (`inFinalizer!()` and some `miniFormatFakeAttributes`). When modding `needsCodegen()` to `return true` in all cases, the same 2 symbols remain undefined (and many more). So the problem for these lie somewhere else; apparently they are not appended to the root module's members and so no codegen candidates.